crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+    { 0, NULL, 0, 0, 0 }, /* treat 0 as error so nothing here */
+    { OSSL_HPKE_AEAD_ID_AES_GCM_128, LN_aes_128_gcm, 16, 16, 12 },
+    { OSSL_HPKE_AEAD_ID_AES_GCM_256, LN_aes_256_gcm, 16, 32, 12 },
+#ifndef OPENSSL_NO_CHACHA20
+# ifndef OPENSSL_NO_POLY1305
+    { OSSL_HPKE_AEAD_ID_CHACHA_POLY1305, LN_chacha20_poly1305, 16, 32, 12 },
+# endif
+    { OSSL_HPKE_AEAD_ID_EXPORTONLY, LN_aes_128_gcm, 16, 16, 12 }
+#endif
+};
+
+/*
+ * @brief info about a KEM
+ */
+typedef struct {
+    uint16_t       kem_id; /**< code point for key encipherment method */
+    const char     *keytype; /**< string form of algtype "EC"/"X25519"/"X448" */
+    const char     *groupname; /**< string form of EC group for NIST curves  */
+    int            groupid; /**< NID of KEM */
+    const char     *mdname; /**< hash alg name for the HKDF */
+    size_t         Nsecret; /**< size of secrets */
+    size_t         Nenc; /**< length of encapsulated key */
+    size_t         Npk; /**< length of public key */
+    size_t         Npriv; /**< length of raw private key */
+} hpke_kem_info_t;
+
+/*
+ * @brief table of KEMs
+ */
+static hpke_kem_info_t hpke_kem_tab[] = {
+    { 0, NULL, NULL, 0, NULL, 0, 0, 0 }, /* treat 0 as error so nowt here */
+    { OSSL_HPKE_KEM_ID_P256, "EC", OSSL_HPKE_KEMSTR_P256, NID_X9_62_prime256v1,
+      LN_sha256, 32, 65, 65, 32 },
+    { OSSL_HPKE_KEM_ID_P384, "EC", OSSL_HPKE_KEMSTR_P384, NID_secp384r1,
+      LN_sha384, 48, 97, 97, 48 },
+    { OSSL_HPKE_KEM_ID_P521, "EC", OSSL_HPKE_KEMSTR_P521, NID_secp521r1,
+      LN_sha512, 64, 133, 133, 66 },
+    { OSSL_HPKE_KEM_ID_X25519, OSSL_HPKE_KEMSTR_X25519, NULL, NID_X25519,
+      LN_sha256, 32, 32, 32, 32 },
+    { OSSL_HPKE_KEM_ID_X448, OSSL_HPKE_KEMSTR_X448, NULL, NID_X448,
+      LN_sha512, 64, 56, 56, 56 }
+};
+
+/*
+ * @brief info about a KDF
+ */
+typedef struct {
+    uint16_t       kdf_id; /**< code point for KDF */
+    const char     *mdname; /**< hash alg name for the HKDF */
+    size_t         Nh; /**< length of hash/extract output */
+} hpke_kdf_info_t;
+
+/*
+ * @brief table of KDFs
+ */
+static hpke_kdf_info_t hpke_kdf_tab[] = {
+    { 0, NULL, 0 }, /* keep indexing correct */
+    { OSSL_HPKE_KDF_ID_HKDF_SHA256, LN_sha256, SHA256_DIGEST_LENGTH },
+    { OSSL_HPKE_KDF_ID_HKDF_SHA384, LN_sha384, SHA384_DIGEST_LENGTH },
+    { OSSL_HPKE_KDF_ID_HKDF_SHA512, LN_sha512, SHA512_DIGEST_LENGTH }
+};
+
+/**
+ * @brief sender or receiver context
+ */
+struct ossl_hpke_ctx_st
+{
+    OSSL_LIB_CTX *libctx; /**< library context */
+    char *propq; /**< properties */
+    int mode; /**< HPKE mode */
+    OSSL_HPKE_SUITE suite; /**< suite */
+    uint64_t seq; /**< sequence number */
+    unsigned char *shared_secret;
+    size_t shared_secretlen;
+    unsigned char *exportersec; /**< exporter secret */
+    size_t exporterseclen;
+    char *pskid; /**< PSK stuff */
+    unsigned char *psk;
+    size_t psklen;
+    EVP_PKEY *senderpriv; /**< sender's ephemeral private key */
```

Comment:
I think the senders private key should not be exposed here... The ephemeral key is generated  and used inside the KEM code only. Only the public key is returned from the KEM for use by the non sender.

Reply:
We dont need it for test vectors because the settable IKM parameter can be used for this purpose.

Reply:
I've taken that out with the "legacy" APIs for now at least.